### ISSUE TITLE
[Question] How to workaround Google rate limiting?

### ISSUE THREAD

**Issue by longnd** [hsnbd/data-scraping#24]
The application hasn't implemented any workaround Google rate limiting. So, in case there are many scarping requests, they will likely blocked by Google.

How would you resolve that issue?

**Comment by hsnbd** [comment]
It could be done by proxy rotation and setting a minimum delay time for every task

**Comment by hsnbd** [comment]
I tried proxy rotation to resolve that issue, but I couldn't find any valid proxy IP to use.

**Comment by hsnbd** [comment]
> another simple solution could be rotating the user's agents used in the request, it makes the requests looks to come from different devices. A simple CURL command would demonstrate that:
> 
> ```shell
> curl -A "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3" https://www.google.com/search\?q\=golang
> ```

I agreed.

**Comment by hsnbd** [comment]
> doesn't give much details on how it looks, would you add a delay when the puppeteer performing the search action, e.g. force it to wait? Depending on the actual implementation, it could be a code smell or not.

In my current solution, I prefetch one message from RabbitMQ at a time and then run Puppeteer. The processing time is already substantial, and there is ample delay by the time a new message enters. 
I am considering if using Axios, which runs quickly and has less delay between two scrapings.

**Comment by longnd** [comment]
> and there is ample delay by the time a new message enters

I assume what you meant was that the delay is the time processing a message. And there is only one message processed at a time, so the next message entered will need to wait. 

If it is the case, it does not give more clarity on the proposed solution for improvement you mentioned:
> setting a minimum delay time for every task

IMO, there are some solutions, including https://github.com/rabbitmq/rabbitmq-delayed-message-exchange

Besides that, the limitations of the approach using Puppeteer were discussed in another issue so we can close this one. Thank you for the response :)
